odoo/cli/obfuscate.py

```diff
@@ -0,0 +1,187 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import odoo
+import sys
+import optparse
+import logging
+
+from pathlib import Path
+from . import Command
+
+_logger = logging.getLogger(__name__)
+
+class Obfuscate(Command):
+    """Obfuscate datas in a given odoo database"""
+    def __init__(self):
+        super().__init__()
+        self.cr = None
+
+
+    def begin(self):
+        if not self.cr:
+            raise Exception("No database connection")
+        self.cr.execute("begin work")
+        self.cr.execute("CREATE EXTENSION IF NOT EXISTS pgcrypto")
+
+    def commit(self):
+        if not self.cr:
+            raise Exception("No database connection")
+        self.cr.commit()
+
+    def rollback(self):
+        if not self.cr:
+            raise Exception("No database connection")
+        self.cr.rollback()
+
+    def set_pwd(self, pwd):
+        """Set password to cypher/uncypher datas"""
+        if not self.cr:
+            raise Exception("No database connection")
+        self.cr.execute("INSERT INTO ir_config_parameter (key, value) VALUES ('odoo_cyph_pwd', 'odoo_cyph_'||encode(pgp_sym_encrypt(%s, %s), 'base64')) ON CONFLICT(key) DO NOTHING", [pwd, pwd])
+
+    def check_pwd(self, pwd):
+        """If password is set, check if it's valid"""
+        if not self.cr:
+            raise Exception("No database connection")
+        uncypher_pwd = self.uncypher_string('value', pwd)
+
+        try:
+            qry = "SELECT " + uncypher_pwd[0] + " FROM ir_config_parameter WHERE key='odoo_cyph_pwd'"
+            self.cr.execute(qry, uncypher_pwd[1])
+            if self.cr.rowcount == 0 or (self.cr.rowcount == 1 and self.cr.fetchone()[0] == pwd):
+                return True
+        except Exception as e:
+            _logger.error("Error checking password: %s", e)
+        return False
+    
+    def clear_pwd(self):
+        """Unset password to cypher/uncypher datas"""
+        if not self.cr:
+            raise Exception("No database connection")
+        self.cr.execute("DELETE FROM ir_config_parameter WHERE key='odoo_cyph_pwd' ")
+
+    def cypher_string(self, field, pwd):
+        # don't double cypher fields
+        return [
+            f"""CASE WHEN starts_with("{field}", 'odoo_cyph_') THEN "{field}" ELSE 'odoo_cyph_'||encode(pgp_sym_encrypt("{field}", %s), 'base64') END""",
+            [ pwd ]
+        ]
+    
+    def uncypher_string(self, field, pwd):
+        return [
+            f"""CASE WHEN starts_with("{field}", 'odoo_cyph_') THEN pgp_sym_decrypt(decode(substring("{field}", 11)::text, 'base64'), %s) ELSE "{field}" END""",
+            [ pwd ]
+        ]
```

Comment:
Wouldn't it be easier if it only returns the case statement, but with a named placeholder. Then, it will be the caller that will explictly pass the `pwd` to `cr.execute`: `cr.execute(query, {"pwd": ...})`
```suggestion
        return 
            f"""CASE WHEN starts_with("{field}", 'odoo_cyph_') THEN pgp_sym_decrypt(decode(substring("{field}", 11)::text, 'base64'), %(pwd)s) ELSE "{field}" END"""
```

(to mix with other suggestion about `psycopg2.sql.SQL`)